src/cc/BPFTable.h

```diff
@@ -68,11 +68,104 @@ class BPFTableBase {
   size_t capacity_;
 };
 
+template <class ValueType>
+class BPFArrayTable : protected BPFTableBase<int, ValueType> {
+private:
+  // this class is used to allow updating a table element using the
+  // table[index] = value syntax
+  class BPFArrayTableWriter {
+    public:
+      BPFArrayTableWriter(BPFArrayTable<ValueType>& t, const int& i):
+      table(t), index(i) {}
+      BPFArrayTableWriter& operator=(ValueType const& rhs) {
+        table.update_value(index, rhs);
+        return *this;
+      }
+      operator ValueType const() {
+        return table.get_value(index);
+      }
+    private:
+      BPFArrayTable<ValueType>& table;
+      const int& index;
+  };
+
+public:
+  BPFArrayTable(BPFModule* bpf_module, const std::string& name)
+      : BPFTableBase<int, ValueType>(bpf_module, name) {
+    int type = bpf_module->table_type(name);
+    if (type != BPF_MAP_TYPE_ARRAY &&
+        type != BPF_MAP_TYPE_PERCPU_ARRAY)
+      throw std::invalid_argument("Table '" + name + "' is not an array table");
+  }
+
+  ValueType get_value(const int& index) {
+    ValueType res;
+    if (!this->lookup(const_cast<int*>(&index), &res))
+      throw std::invalid_argument("index out of range");
+    return res;
+  }
+
+  // updates an element
+  void update_value(const int& index, const ValueType& value) {
+    if (!this->update(const_cast<int*>(&index), const_cast<ValueType*>(&value)))
+      throw std::invalid_argument("Error updating element");
```

Comment:
I based my code on the existing one, IMHO opinion we should continue to use exceptions, for example these are some issues if we only use return codes:
- In case of constructors we should implement a static method that instantiates objects as suggested by @4ast 
- It is impossible to use return codes with the [] operator
- It is impossible to return multiple values in a c++ function, then the return value should be passed as a parameter of the function. 

Of course these can be solved, but maybe the pain of doing that is more that remaining with the exception-based approach.

Said that, is it worthy to avoid using exceptions? Am I missing something about using exceptions?